### ISSUE TITLE
Hard to catch error on proxying in Node.js 10

### ISSUE THREAD

**Issue by kmannislands** [webpack/webpack-dev-server#1642]
* Operating System: macOs 10.14.1
* Node Version: 10.15.0
* NPM Version: yarn 1.13.0
* webpack Version: 4.23.1
* webpack-dev-server Version: 3.1.14

- [X] This is a **bug**
- [ ] This is a **modification** request

### Code
The bug I'm reporting here showed up after switching to Node js v10.x. I believe there are differences in how the `net` core package raises exceptions/policy toward unhandled error events.

To repro, you need to proxy a connection that uses WebSockets, allow it to connect using chrome, then restart the target server of the proxy, for example with Nodemon. Want to avoid creating a minimal repo of this if possible since it's a bit comple and there seems to be a nice trail of ticket/changelogs.

This will cause the `WDS` process to die due to unhandled error event from `ECONNRESET` with the following stack trace:

```
HPM] Error occurred while trying to proxy request /foo from localhost:9090 to https://localhost:3333 (ECONNREFUSED) (https://nodejs.org/api/errors.html#errors_common_system_errors)
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TLSWrap.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at Object.apply (/Users/kieran/work/Atlas/ui/node_modules/harmony-reflect/reflect.js:2064:37)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```
Before upgrading to node js 10, I believe something similar was happening but the exception wasn't fatal.

After some debugging/ticket searching it seems that what changed is that it is now an error for the https module to encounter an exception without a registered handler, which impacts the behavior of your dependency `node-http-proxy`, [like this ticket describes](https://github.com/nodejitsu/node-http-proxy/issues/1286).

Debugging the error, I discovered that WDS `proxy` config is really `http-proxy-middleware` config, so it is possible to suppress the error and keep the server up by modifying config by adding [an onError handler to proxy config](https://github.com/chimurai/http-proxy-middleware#http-proxy-events) for future readers, however this was fairly difficult to debug and ideally this package would register some sort of handler for this exception so that it isn't fatal.

Willing to submit a PR, the change is very straight-forward. Involve modifying [this code](https://github.com/webpack/webpack-dev-server/blob/519b7036191702992bd0ea5f037db5ac605a8ca1/lib/Server.js#L630):
```js
  // Proxy websockets without the initial http request
  // https://github.com/chimurai/http-proxy-middleware#external-websocket-upgrade
  websocketProxies.forEach(function (wsProxy) {
    this.listeningApp.on('upgrade', wsProxy.upgrade);
  }, this);
```
To be like: 
```js
// Proxy websockets without the initial http request
  // https://github.com/chimurai/http-proxy-middleware#external-websocket-upgrade
  websocketProxies.forEach(function (wsProxy) {
    this.listeningApp.on('upgrade', (req, socket, ...args) => {
      socket.on('error', (err) => console.error(err));
      return wsProxy.upgrade(req, socket, ...args);
    };
  }, this);
```
### Expected Behavior
The server should stay alive when a client (on either side of the proxy) disconnects.

### Actual Behavior
Server dies due to no handler for `ECONNRESET`

### For Bugs; How can we reproduce the behavior?
Use Node 10.

**Comment by evilebottnawi** [comment]
@kmannislands Can you create minimum reproducible test repo, it is hard to fix without this and maybe never will be fixed without this

**Comment by evilebottnawi** [comment]
@kmannislands Fatal error should be don't crash webpack-dev-server so we will accept a PR, but we need test case or minimum reproducible test repo to unsure we fix this problem and avoid regressions in future

**Comment by evilebottnawi** [comment]
@kmannislands we need minimum reproducible test repo for manual testing (hard to testing), it is increase speed for solving problem

**Comment by mjrussell** [comment]
Can confirm we are also experiencing this error proxying websockets when the server doesn't gracefully respond to the connect or it closes abruptly. Only happens on Node 10.x

**Comment by evilebottnawi** [comment]
@mjrussell can you create minimum reproducible test repo?

**Comment by evilebottnawi** [comment]
Friendly ping guys

**Comment by cha0s** [comment]
Yes it's still an issue.

No I can't produce a minimal repro, concurrency is hard.

Please just fix it :heart:

**Comment by evilebottnawi** [comment]
@cha0s it is hard to fix something in anything, can you provide screen this about problem?

**Comment by mathieumg** [comment]
@evilebottnawi It seems @kmannislands provided the console output of the error, so pretty much equivalent to a screenshot. Basically, you start the Webpack bundler with proxying, and if any client disconnects it crashes the whole process for the reason mentioned in the original issue, above:

<img width="1336" alt="Screen Shot 2019-04-23 at 2 00 32 PM" src="https://user-images.githubusercontent.com/1624812/56604626-698d9480-65d0-11e9-935e-5ade2c21873e.png">

Happens on Node v10.x.

**Comment by hiroppy** [comment]
@mathieumg Only Node v10.x? Node.js v12 has been released today so I want to know what occurs using v12. Thanks.

**Comment by mathieumg** [comment]
@hiroppy I presume it will be the same on v12, but I can install it and test it too. Either way, this still needs to be fixed for v10.

**Comment by mathieumg** [comment]
@hiroppy A separate bug prevents me from testing it in v12, but I'm not on the latest version of the Dev Server either:

```
Starting the development server...

internal/buffer.js:928
class FastBuffer extends Uint8Array {}
^

RangeError: Invalid typed array length: -4095
    at new Uint8Array (<anonymous>)
    at new FastBuffer (internal/buffer.js:928:1)
    at Handle.onStreamRead [as onread] (internal/stream_base_commons.js:165:17)
    at Stream.<anonymous> (/Users/mathieumg/project/frontend/node_modules/handle-thing/lib/handle.js:120:12)
    at Stream.emit (events.js:201:15)
    at endReadableNT (/Users/mathieumg/project/frontend/node_modules/readable-stream/lib/_stream_readable.js:1010:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:17)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by hiroppy** [comment]
see https://github.com/nodejs/node/issues/24097

**Comment by hiroppy** [comment]
@mathieumg Could you submit a reproducible repo?

**Comment by mathieumg** [comment]
> Could you submit a reproducible repo?

Yes I can try to muster up a repro repo when I have a spare second sometime this week!

**Comment by dhardtke** [comment]
I tried reproducing this error in a minimal repo, but I can't. Tried with NodeJS 10, 11 and 12. NodeJS 12 doesn't work with WDS though (getting a weird error message different than the one posted above - maybe because I'm on windows).

Maybe you guys can use the repo I created. I have definitely seen this behavior in our project using NodeJS 11.x as well:

https://github.com/dhardtke/wds-ws-proxy-crash

**Comment by kmannislands** [comment]
IIRC, this also may have depended on a specific chrome version. Perhaps chrome has changed its behavior again.

To clarify my original ticket, this is non-blocking because WDS allows users to pass through config to the proxy package. So if anyone reading this is blocked, all you need to do is add an error handler on upgrade to the WDS config in your webpack config:

```js
{
  // ... other config
  devServer: {
    proxy: {
      '/your_path': {
        target: 'https://localhost:XXX',
        ws: true,

        // Add this:
        onError(err) {
          console.log('Suppressing WDS proxy upgrade error:', err);
        },
      },
    },
  },
}
```

**Comment by evilebottnawi** [comment]
ho, i think we should add logger by default as described above, feel free to send a PR, it is easy to fix

**Comment by markmcdowell** [comment]
This has just started happening to me but I'm unsure of the change that caused it.

- Operating System: macOs 10.14.5
- Node Version: 10.15.3
- NPM Version: 6.9.0
- webpack Version: 4.31.0
- webpack-dev-server Version: 3.3.1

The onError addition above doesn't make a difference I still get the crash with:

```
events.js:174                                                                       
       throw er; // Unhandled 'error' event
       ^
Error: read ECONNRESET
     at TCP.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
     at emitErrorNT (internal/streams/destroy.js:82:8)
     at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
     at process._tickCallback (internal/process/next_tick.js:63:19)
```

It happens at the point I try to open a websocket through the proxy (it doesn't always happen):
```
proxy: {
            "/api": {
                target: "http://localhost:8181",
            },
            '/api/realtime': {
                target: "ws://localhost:8181",
                ws: true
            },
            "/auth": {
                target: "http://localhost:8181",
            },
        }
```

**Comment by evilebottnawi** [comment]
@markmcdowell can you create minimum reproducible test repo?

**Comment by scipper** [comment]
I got the same error. What actually helped me out was:
```
process.on('uncaughtException', function (err) {
  console.error(err.stack);
  console.log("Node NOT Exiting...");
});
```
Found the solution on [StackOverflow](https://stackoverflow.com/a/34186272/8569519). Thanks to Ross the Boss.

**Comment by evilebottnawi** [comment]
Friendly ping, somebody still have the problem? Please create reproducible test repo, thanks

**Comment by alexkuz** [comment]
@evilebottnawi I tried to force install latest `webpack-dev-server` version (I'm using `create-react-app`) and it seems that error is still there.
I might have a try in making an example (it needs some setup work), although personally @scipper solution is fine for me.

**Comment by evilebottnawi** [comment]
@alexkuz will be great if you help to reproduce the problem, because @scipper solution is not good, I think we don't catch some errors, so we need fix it

**Comment by alexkuz** [comment]
@evilebottnawi ok, here it is (based on CRA):

https://github.com/alexkuz/wds-ws-proxy-example

To reproduce:
- install deps with `yarn install`
- run server with `node src/server.js`
- open dev app with `yarn start`

Server returns `502` error instead of proper WS response (turns out that's what Nginx does on my server when I'm for example restarting container with my actual WS server), which leads to `ECONNRESET` error.

**Comment by evilebottnawi** [comment]
@alexkuz your problem is not related to the original problem :confused: What is use case? Why do you use it configuration? You can send a fix, but it is not high priority, because it is edge case.

**Comment by alexkuz** [comment]
> your problem is not related to the original problem 😕

you may be right, I haven't read original issue very carefully tbh, sorry 😬 The error looks very similar though so it might be related.

> Why do not use real host for ws instead proxing?

You now what, I just tried to connect directly and it works 😅 I configured this long time ago and not sure why (or if) it really didn't work before. Something with CORS, maybe?

Still, I think catching that 502 error wouldn't hurt, but yeah, it's probably a low priority issue.

**Comment by alexkuz** [comment]
Ah, I think the idea was to not use server url explicitly in source code, it was easier to setup proxy than providing url via env variable or something.

**Comment by evilebottnawi** [comment]
> Something with CORS, maybe?

I think not, a problem in `ws` implementation, i think you can look at here https://github.com/websockets/ws/issues/1692

@alexkuz Yes, anyway feel free to send a PR, i think it should be easy to fix :+1:

**Comment by rightaway** [comment]
Issue still exists in Node 13. And isn't specific to websockets, happens even without websocket use.

**Comment by BeaudanBrown** [comment]
This is not really a repro but I believe I have the same issue and it maybe it will be helpful. Here is the repo I'm just trying to start a new project in

https://github.com/BeaudanBrown/boardgames

To build probably need
`npm i -g spago` first, and then
```
npm install
spago install
spago build
npm run webpack:server
```
This is running the command
`webpack-dev-server --progress --inline --hot`

If I then try to save the file, seemingly at random it will fail with the following log:

```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/home/beau/.config/nvm/versions/node/v13.13.0/bin/node',
1 verbose cli   '/home/beau/.config/nvm/versions/node/v13.13.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'webpack:server'
1 verbose cli ]
2 info using npm@6.14.4
3 info using node@v13.13.0
4 verbose run-script [ 'prewebpack:server', 'webpack:server', 'postwebpack:server' ]
5 info lifecycle boardgames@1.0.0~prewebpack:server: boardgames@1.0.0
6 info lifecycle boardgames@1.0.0~webpack:server: boardgames@1.0.0
7 verbose lifecycle boardgames@1.0.0~webpack:server: unsafe-perm in lifecycle true
8 verbose lifecycle boardgames@1.0.0~webpack:server: PATH: <REDACTED>
9 verbose lifecycle boardgames@1.0.0~webpack:server: CWD: /home/beau/work/purescript/boardgames
10 silly lifecycle boardgames@1.0.0~webpack:server: Args: [ '-c', 'webpack-dev-server --progress --inline --hot' ]
11 silly lifecycle boardgames@1.0.0~webpack:server: Returned: code: 1  signal: null
12 info lifecycle boardgames@1.0.0~webpack:server: Failed to exec webpack:server script
13 verbose stack Error: boardgames@1.0.0 webpack:server: `webpack-dev-server --progress --inline --hot`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/home/beau/.config/nvm/versions/node/v13.13.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:315:20)
13 verbose stack     at ChildProcess.<anonymous> (/home/beau/.config/nvm/versions/node/v13.13.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:315:20)
13 verbose stack     at maybeClose (internal/child_process.js:1026:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
14 verbose pkgid boardgames@1.0.0
15 verbose cwd /home/beau/work/purescript/boardgames
16 verbose Linux 5.6.3-arch1-1
17 verbose argv "/home/beau/.config/nvm/versions/node/v13.13.0/bin/node" "/home/beau/.config/nvm/versions/node/v13.13.0/bin/npm" "run" "webpack:server"
18 verbose node v13.13.0
19 verbose npm  v6.14.4
20 error code ELIFECYCLE
21 error errno 1
22 error boardgames@1.0.0 webpack:server: `webpack-dev-server --progress --inline --hot`
22 error Exit status 1
23 error Failed at the boardgames@1.0.0 webpack:server script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Let me know if there's any other information I can provide

**Comment by ueler** [comment]
Is an issue for me as well with proxied websockets to our backend server... this is bad because restarting the backend server always kill webpack dev...

The only solution that worked for us for now is the one catching the expeption (thanks @scipper!)

**Comment by evilebottnawi** [comment]
@ueler Can you create reproducible test repo?

**Comment by rameshsunkara** [comment]
We had the same problem and worked around the problem by downgrading webpack-dev-server to v3.9.0 instead of using v3.11.0

**Comment by evilebottnawi** [comment]
@rameshsunkara Please create reproducible test repo

**Comment by vperron** [comment]
All of my team has had this issue throughout the last year. The fix seems to stand in two lines !
I can't recount how many times the sentence "reproducible test repo" has been said but it is a shrödingbug, very hard to reproduce but yes, it exists and spans at least 3 different node versions.

Why not trying the optimistic merging on good faith ? People have repeatedly said that the fix was working. Just add a long comment to explicit why it is so hard to reproduce, but making people lives easier is a good thing, isn't it ?

**Comment by yoyo837** [comment]
Same issue https://github.com/chimurai/http-proxy-middleware/issues/44#issuecomment-655948463

**Comment by evilebottnawi** [comment]
Somebody can create reproducible test repo or step to reproduce or we will close issue

**Comment by yoyo837** [comment]
It is too hard, it requires a connectable test Websocket server and can be started or stopped.

**Comment by evilebottnawi** [comment]
@yoyo837 Just write your steps

**Comment by cha0s** [comment]
Can you imagine going a year and a half commenting over and over expecting people to be able to reproduce some obscure race condition instead of just taking the first comment which had code completely laid out to fix this? A year and a half ago. 

There's no way you can convince me that all of these comments over the last **year and a half** constituted less effort than just rolling in the fix.

I worked around this long ago by simply avoiding this broken functionality. I feel for anyone still dealing with this *after a year and a half*.

**Comment by yoyo837** [comment]
@evilebottnawi 

proxy config in webpack options:

```
  proxy: {
    '/v3/websocket/socketServer.do': {
      target: 'http://localhost:8080/',
      changeOrigin: true,
      ws: true,
    },
    '/v3': {
      target: 'http://localhost:8080/',
      changeOrigin: true,
    },
  },
```

Steps:
1. Start java or any other server (websocket server)
2. Start dev server (websocket client)
3. Open the browser using the dev URL, and the websocket code in it will connect to the server
4. Shut down the server forcefully, the client attempts to reconnect at 3 second intervals
5. Wait a few minutes or more time
6. The dev server crash

![imange](https://user-images.githubusercontent.com/6134547/87008972-17e22180-c1f7-11ea-8477-f59a00469e39.png)

**Comment by evilebottnawi** [comment]
@cha0s why be toxic? If you read, you will see that there were different problems, I ask this to understand in detail and fix all possible problems, and not just take and close the issue. And it’s very disappointing to hear such words when you try fix and improve your work

@yoannmoinet big thanks!

**Comment by Loonride** [comment]
> /cc @Loonride Maybe you have time to look at this. I will have time only on the next week

Will look later, I'm working on many things

**Comment by evilebottnawi** [comment]
Thanks for helping, no rush

**Comment by cha0s** [comment]
I apologize for my tone. I will just say that it is always possible for perfect to be the enemy of good enough. It's better (for your loving users) to fix and then keep the issue open to perfect the fix if you believe the very first comment's fix is wrong. I haven't seen a good argument for why it is the wrong fix, and I suspect it will be the ultimate resolution.

**Comment by evilebottnawi** [comment]
@cha0s because any line of code should be tested, it can have side effects when you adding lines without test, like regressions, new problems and etc, so I ask developers provide more information for tests

**Comment by ciekawy** [comment]
The case I' facing seems pretty typical. I have Angular client app served by `webpack-dev-server`. I use proxy to expose backend to the client app. Recently I added web socket connection (in particular to handle Apollo GraphQL subscriptions over websockets) which is proxied.
During development on each backend restart the proxy dies due to this issue. After manually applying proposed change from the issue description to `Server.js` the proxy stays alive and client is able to reconnect on subsequent backend restarts.

**Comment by wrench7** [comment]
For anyone still coming across this issue, the initial [fix code](https://github.com/webpack/webpack-dev-server/issues/1642#issue-403313084) didn't work for me. There was a missing parentheses, the correct code is:
```
  websocketProxies.forEach((wsProxy) => {
    this.listeningApp.on('upgrade', (req, socket, ...args) => {
      socket.on('error', (err) => console.error(err));
      return wsProxy.upgrade(req, socket, ...args);
    });
  }, this);
```

**Comment by maknapp** [comment]
This error can be caught using the `onProxyReqWs` event in the config:

```
devServer: {
  proxy: {
    '/my-websocket-path': {
      target: 'wss://my.ip',
      xfwd: true,
      ws: true,
      secure: false,
      onProxyReqWs: (proxyReq, req, socket, options, head) => {
        socket.on('error', function (err) {
          console.warn('Socket error using onProxyReqWs event', err);
        });
      }
    }
  }
}
```

**Comment by ciekawy** [comment]
Awesome workaround! thanks @maknapp for sharing!
> This error can be caught using the `onProxyReqWs` event in the config:

**Comment by maknapp** [comment]
I can reproduce this error consistently by setting the target to nginx to then `proxy_pass` to nowhere (e.g. a destination that is being restarted).

```
events {}

http {
  proxy_http_version 1.1;

  # This is intended to point to nothing so that
  # the request to /ws fails.
  upstream _serverThatDoesNotExist {
    server 127.0.0.1:7070;
  }

  server {
    listen 7071 default;

    root /var/www;

    location /ws {
      proxy_pass http://_serverThatDoesNotExist;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "Upgrade";
    }
  }
}
```

```
devServer: {
  proxy: {
    '/ws': {
      target: 'ws://127.0.0.1:7071',
      xfwd: true,
      ws: true,
      secure: false,
      onProxyReqWs: (proxyReq, req, socket, options, head) => {
        socket.on('error', function (err) {
          console.warn('Socket ERROR using onProxyReqWs', err);
        });
      }
    }
  }
}
```

A websocket repeatedly attempting to connect to the webpack-dev-server will produce the following every time:
```
Error: read ECONNRESET at TLSWrap.onStreamRead (internal/stream_base_commons.js:209:20)
```


If the proxy target is set directly to something that does not exist (e.g. a destination that is being restarted), it will mostly get a connect error (which is caught by default) and sometimes get this same read error. It seems there is some race condition somewhere between connecting, reading, and the default error handling.

**Comment by evilebottnawi** [comment]
@maknapp Maybe do you want to send a fix, we can add default error handler

**Comment by maknapp** [comment]
Should this be fixed on http-proxy-middleware instead of here? It already has a default error handler and seems like the more appropriate place to add another.

**Comment by alexander-akait** [comment]
Yea and no, we have own logget and we need to ability setup output level in this case, so I think we need to setup own logger here

**Comment by mecampbellsoup** [comment]
Hi guys - I have the following webpack config, and intermittently WDS crashes with the below error.

```
  devServer: {
    proxy: {
      '/api': 'http://gatekeeper',
      ws: true,
    },
    contentBase: commonPaths.outputPath,
    compress: true,
    hot: true,
    host: '0.0.0.0',
    public: 'cloud.coreweave.test',
```
(I am running WDS inside of a container, inside of a minikube cluster/host, hence the `public` configuration option.)

```
[gatekeeper-frontend] Version: webpack 4.46.0
[gatekeeper-frontend] Time: 347ms
[gatekeeper-frontend] Built at: 03/04/2021 2:50:54 AM
[gatekeeper-frontend]  1 asset
[gatekeeper-frontend] Entrypoint main = main.js
[gatekeeper-frontend] [./src/App.jsx] 2.97 KiB {main} [built]
[gatekeeper-frontend] [./src/index.jsx] 705 bytes {main} [built]
[gatekeeper-frontend]     + 302 hidden modules
[gatekeeper-frontend] ℹ ｢wdm｣: Compiled successfully.
[gatekeeper-frontend] events.js:291
[gatekeeper-frontend]       throw er; // Unhandled 'error' event
[gatekeeper-frontend]       ^
[gatekeeper-frontend]
[gatekeeper-frontend] Error: read ECONNRESET
[gatekeeper-frontend]     at TCP.onStreamRead (internal/stream_base_commons.js:209:20)
[gatekeeper-frontend] Emitted 'error' event on Socket instance at:
[gatekeeper-frontend]     at emitErrorNT (internal/streams/destroy.js:100:8)
[gatekeeper-frontend]     at emitErrorCloseNT (internal/streams/destroy.js:68:3)
[gatekeeper-frontend]     at processTicksAndRejections (internal/process/task_queues.js:80:21) {
[gatekeeper-frontend]   errno: -104,
[gatekeeper-frontend]   code: 'ECONNRESET',
[gatekeeper-frontend]   syscall: 'read'
[gatekeeper-frontend] }
[gatekeeper-frontend] error Command failed with exit code 1.
```
Is there anything _obviously wrong_ with my configuration here?

**Comment by rafaelsales** [comment]
@mecampbellsoup Have you tried this solution: https://github.com/webpack/webpack-dev-server/issues/1642#issuecomment-733826290 ?

By the way, when you say "intermittently", this should typically occur when the WS proxy connection drops (e.g. when the target proxy server restarts). Networking issues shouldn't be the norm, but could also be the culprit.

Ideally WDS would have this default error handler and we wouldn't have to write a custom one just to ignore the error - in fact, this is an expected error and it should never crash the server.

**Comment by mecampbellsoup** [comment]
> @mecampbellsoup Have you tried this solution: [#1642 (comment)](https://github.com/webpack/webpack-dev-server/issues/1642#issuecomment-733826290) ?
> 
> By the way, when you say "intermittently", this should typically occur when the WS proxy connection drops (e.g. when the target proxy server restarts). Networking issues shouldn't be the norm, but could also be the culprit.
> 
> Ideally WDS would have this default error handler and we wouldn't have to write a custom one just to ignore the error - in fact, this is an expected error and it should never crash the server.

It seems to just happen at random - it seems like perhaps my WS client (in the browser) is disconnecting for some reason, kicking off the following state changes in my WDS server (the WS proxy backend):

```
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     Error     1          2m36s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     CrashLoopBackOff   1          2m41s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     Running            2          2m54s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              1/1     Running            2          2m59s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     Error              2          3m57s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     CrashLoopBackOff   2          4m1s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              0/1     Running            3          4m29s
cloud          gatekeeper-frontend-66ffbc7dc4-vjkmw                              1/1     Running            3          4m39s
```
After restarting, things generally work again for a while, until the next occurrence just a few short minutes later...

I added `onError` and `onProxyReqWs` handlers to my `devServer.proxy` config but the errors don't seem to be caught.

**Comment by alexander-akait** [comment]
@mecampbellsoup Maybe you can create simple reproducible test repo, I will look, I think we fix it in master, jsut want to check

**Comment by mecampbellsoup** [comment]
> @mecampbellsoup Maybe you can create simple reproducible test repo, I will look, I think we fix it in master, jsut want to check

When I run this app outside of minikube, I don't seem to have any WS connection issue or disconnects, so creating a reproducible test repo is going to require a docker compose or minikube deploy script... if you need this I'll need some time.

**Comment by alexander-akait** [comment]
@mecampbellsoup Yep, will be great

**Comment by mecampbellsoup** [comment]
> @mecampbellsoup Yep, will be great

@alexander-akait @rafaelsales, as with so many things, this turned out to be a user error 😢 

In short, my WDS proxy config was invalid. Compare the following two:

### Bad

```
proxy: {
      '/api': process.env.PROXY_URI,
      ws: true,
      onError: err => {
        console.log('Suppressing WDS proxy upgrade error:', err);
      },
    },
```

### Good

```
proxy: {
      '/api': {
        target: process.env.PROXY_URI,
        ws: true,
        onError: err => {
          console.log('Suppressing WDS proxy upgrade error:', err);
        },
      },
    },
```
I blame JS/WDS for letting me pass an *invalid but correct-looking* proxy configuration object!

**Comment by alexander-akait** [comment]
Unfortunately, we cannot validate invalid parameters for 3d party modules :disappointed: But I will look how we can import it, thanks for reporting

**Comment by alexander-akait** [comment]
After spending 5 hours I found how to reproduce the problem

**Comment by yoyo837** [comment]
Thanks for your hard works from community.

**Comment by chimurai** [comment]
Hi @alexander-akait 

I just published [http-proxy-middleware@1.1.0](https://github.com/chimurai/http-proxy-middleware/releases/tag/v1.1.0) which fixed a graceful shutdown issue. Curious if it fixes this issue.

If not I'm interested in the reproducible setup. Maybe it's something HPM can fix (if issie can be reproduced)
Spend some hours on it without luck...

**Comment by alexander-akait** [comment]
@chimurai Yep, it is interesting problem, I reproduce it using IE11/IE10, proxy websockets and sockjs, I don't know why it happens, anyway I try to reproduce it again and put feedback on the next verison

**Comment by simonh1000** [comment]
Just hit this as well. 

Operating System: macOs 11.2.3
Node Version: v14.15.3
NPM Version: 7.5.2
"webpack": "^4.44.2",
"webpack-dev-server": "^3.11.0",

I have this proxying
                {
                    // ws:localhost:3002 --> ws:localhost:8000
                    context: ["/a/"],
                    target: ampTarget,
                    pathRewrite: { "^/a": "" },
                    ws: true,
                    headers: {
                        "X-Remix-Endpoint": `http://localhost:${port}/a`,
                    },
                },

and I hit the crash immediately when I try e.g. http://192.168.0.148:3002/... in my browser

**Comment by mlilback** [comment]
Here is a reproducible test case. using node v16.3.0, Chrome 91.0.4472.114 on macOS 11.4.

I'm using Vue, so I just followed the webpack documentation to create a project. You don't have to be running an actual server listening: the process never gets that far for me.

**Comment by alexander-akait** [comment]
@mlilback sorry for delay, thanks, I will look at this in near future

**Comment by alexander-akait** [comment]
@mlilback Do you have crashed dev server in your example? Because I can't reproduce due I don't have ability to do request for `http://proxyman.local:3145/ `

**Comment by veloek** [comment]
In our case this issue is very easy to reproduce. We use webpack-dev-server to proxy websocket requests to an application running in IIS. To reproduce this issue we simply recycle the app pool of the site.

Have used patch-package to apply @kmannislands' fix, but now I see the problem can be resolved with @maknapp's config suggestion which feels like a cleaner workaround. Would like to see it fixed properly though :slightly_smiling_face:

**Comment by alexander-akait** [comment]
@veloek Can you provide steps to reproduce or reproducible test repo? To be honestly here three different issues

**Comment by veloek** [comment]
@alexander-akait In our scenario there are unfortunately too many moving parts for me to easily create a test repo, but I guess any setup where you kill the websocket server while a client is connected should trigger the same issue. I sadly don't have the time to dive deeper into it as long as we have a workaround.

**Comment by alexander-akait** [comment]
@veloek maybe you can provide full stack trace of the error?

**Comment by veloek** [comment]
@alexander-akait This is the error message printed to the console when I recycle the IIS app pool causing WDS to crash:
```
events.js:352
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:209:20)
Emitted 'error' event on Socket instance at:
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:82:21) {
  errno: -4077,
  code: 'ECONNRESET',
  syscall: 'read'
}
```

**Comment by alexander-akait** [comment]
@0x53A Can you provide example to reproduce? What is `webpack-dev-server` version? I think it should be solved for v4, if not, we need investigate, fix should be easy, just need understand where it happens

**Comment by 0x53A** [comment]
@alexander-akait 

I'm was on

```
webpack-dev-server: 4.2.0
http-proxy-middleware: 2.0.1

nodejs: v16.8.0
```

and updated to ``webpack-dev-server: 4.3.1``.

With the latest version it does seem to work, thank you!

I didn't update before complaining because the issue was still open, sorry.

**Comment by markusschaber** [comment]
> @alexander-akait
>
> I think the real fix would be [nodejs/node#36111](https://github.com/nodejs/node/pull/36111).

I'm not sure about that - I run into this error without any TLS involved (as everything is on localhost):

```
√ Compiled successfully.
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:4200/api/JsonRpc to undefined [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)  
node:events:498
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (node:internal/stream_base_commons:217:20)
Emitted 'error' event on Socket instance at:
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -4077,
  code: 'ECONNRESET',
  syscall: 'read'
}

```

**Comment by Delagen** [comment]
Issue about 3 years, but still patch lives in my sources in postinstall action
```diff
diff --git a/node_modules/webpack-dev-server/lib/Server.js b/node_modules/webpack-dev-server/lib/Server.js
--- a/node_modules/webpack-dev-server/lib/Server.js
+++ b/node_modules/webpack-dev-server/lib/Server.js
@@ -1840,7 +1840,10 @@ class Server {
       (this.server).on(
         "upgrade",
         /** @type {RequestHandler & { upgrade: NonNullable<RequestHandler["upgrade"]> }} */
-        (webSocketProxy).upgrade
+        (req, socket, ...args) => {
+            socket.on('error', (err) => console.error(err));
+            return webSocketProxy.upgrade(req, socket, ...args);
+        }
       );
     }, this);
   }

```

**Comment by alexander-akait** [comment]
@Delagen Can you send a PR? Will be great to have test case, I tried to reproduce but no luck

**Comment by chimurai** [comment]
Hi @alexander-akait.

Might have found the issue while working on the next version of http-proxy-middleware.

I noticed `http-proxy` has an undocumented `econnreset` event.

Added a [default handler](https://github.com/chimurai/http-proxy-middleware/pull/759/files) to `econnreset`; Hopefully this will prevent the server from crashing. 🤞
Haven't been able to replicate this bug as well.

Just published [http-proxy-middleware@2.0.5](https://github.com/chimurai/http-proxy-middleware/releases/tag/v2.0.5)

Curious to hear if `v2.0.5` fixes this gnarly bug

**Comment by alexander-akait** [comment]
@chimurai Big thanks

@Delagen Can you test it without hack and with a new version?

**Comment by maknapp** [comment]
I can confirm this fixes the issue with webpack-dev-server@4.7.4 and http-proxy-middleware@2.0.5

**Comment by Delagen** [comment]
@chimurai Thanks for fast fix. It works now.

**Comment by alexander-akait** [comment]
Great! I think we can close it, anyway if you faced with this issue - feel free to open an issue, hope we covered all cases

**Comment by chimurai** [comment]
Glad to hear the fix is working.

Thanks @alexander-akait for helping out in this thread. 💪

**Comment by maknapp** [comment]
It seems I spoke too soon. I am still getting the error. Somehow the ECONNRESET error is thrown without being caught. I see the error handler in http-proxy-middleware being added, but it does not catch anything.

**Comment by maknapp** [comment]
@chimurai The `econnreset` event appears to be only for http and not websocket. The uncaught error causing the crash when websockets are involved is an "error" event, not an "econnreset" event. Looking at https://github.com/http-party/node-http-proxy/blob/master/lib/http-proxy/passes/ws-incoming.js, the uncaught "error" event is coming from `socket`. There is a `proxySocket.on("error")` that re-emits to the "error" handler in http-proxy-middleware, but there is none for `socket.on("error")`. 

The only way I see to catch the error is to add a default `proxyReqWs` handler that then adds an error handler to `socket`. The problem with that is someone with a proxyReqWs handler will not know to handle socket errors.

**Comment by maknapp** [comment]
In http-proxy, there is an error handler on `socket` that is added only after the websocket upgrade. The ECONNRESET error causing the crash occurs _before_ the upgrade (since there is no destination) and throws an error that nobody catches. This seems to be fixed by https://github.com/http-party/node-http-proxy/pull/1439, but it looks like there is nobody there to merge it.

The following in http-proxy-middleware fixes all crashing for me and allows users to still have an `onProxyReqWs` handler:
```js
proxy.on('econnreset', (err, req, res, target) => {
    logger.error(`[HPM] ECONNRESET: %s`, err.message);
});
proxy.on('proxyReqWs', (proxyReq, req, socket, options, head) => {
    socket.on('error', (error) => {
        logger.error(`[HPM] WebSocket error: %s`, error.message);
    });
});
```

Note that I changed the econnreset handler to log `err.message`, not `err` since it printed the object.

**Comment by chimurai** [comment]
Appreciate for your feedback @maknapp!

Published `http-proxy-middleware@2.0.6` with your suggestion.

Let me know if this fix helps.

Bit challenging without  a reproduction of the bug 😬

**Comment by veloek** [comment]
@chimurai I can confirm that version 2.0.6 solves the issue. It used to crash for me every time the app pool recycled (IIS), but now it just logs this and keeps running:
```
[webpack-dev-server] [HPM] WebSocket error: Error: read ECONNRESET
    at TCP.onStreamRead (node:internal/stream_base_commons:217:20) {
  errno: -4077,
  code: 'ECONNRESET',
  syscall: 'read'
}
```
So now I can remove the workaround, great work! 🙂

**Comment by maknapp** [comment]
v2.0.6 is working for me as well. Thanks @chimurai!

**Comment by VitaliyR** [comment]
Is there any way to silent exactly these messages?

**Comment by alexander-akait** [comment]
You can setup logging for http-proxy-middleware, there are options for it
